### ISSUE TITLE
PHPMailer assumes that sendmail_path always contains the word 'sendmail'

### ISSUE THREAD

**Issue by coldscientist** [PHPMailer/PHPMailer#2181]
# Problem description

The PHPMailer assume that the PHP directive ``sendmail_path`` always contains ``sendmail`` into it's path through the ``isSendmail()`` function (PHPMailer.php), but it's not always true.

The main objective of ``sendmail_path`` directive is exactly specifying a third party executable and/or additional command line parameters arguments that PHP will use to send mails through the built-in ``mail()`` function. 

Not all servers have a full-blown Postfix instance to send mails. In my case, I have a localhost only server that isn't available through the Internet and could not send mails by itself, and my e-mail provider doesn't allow Relay, so I need to use a fake MTA (like [msmtp](https://wiki.archlinux.org/index.php/msmtp)) to send mail to external e-mail address.

To [send mail with PHP using msmtp](https://wiki.archlinux.org/index.php/msmtp#Send_mail_with_PHP_using_msmtp), I must change the ``sendmail_path`` directive to use msmtp instead of ``sendmail``:

```
sendmail_path = "/bin/msmtp -C /etc/msmtprc -t  --read-envelope-from"
```

But if I try to do it, PHPMailer doesn't work. I'm using GLPI and detected that uses PHPMailer to send mails instead of the ``mail()`` function, even when selecting to use ``Mail`` instead of ``PHPMailer``, but when PHPMailer is called, it wasn't sending mails, but it was working fine through shell and when sending it directly through PHP ``mail()`` function:

```
sudo -u web1 -s php -r 'mail("eduardo.mozart@example.com","Testing php -v ".phpversion(),"php on ".gethostname(),"From: GLPI Project <glpi@example.com>");' -c /var/www/conf/web1/php.ini
```

It was working fine (I'm using ISPConfig to host my local webserver, but I don't believe it's relevant) with the command above, so I decided to investigate PHPMailer source code and found the ``isSendmail()`` function (PHPMailer.php), I commented it to always return ``$this->Sendmail = $ini_sendmail_path;`` and voila, the mails were sent normally through GLPI/PHPMailer.

I believe that the function ``isSendmail()`` was added to the code to avoid breaking the sendmail functionality because of user misconfiguration, but it limits the purpose of ``sendmail_path`` PHP directive by itself. It was a bit tricky to debug this issue because it was working fine through Shell and calling ``mail()`` function directly.

Some distros provides packages to replace the Postfix ``sendmail`` with a fake one (like ``msmtp-mta`` package available at Arch repositories), so the msmtp is called when sendmail is executed and the ``stripos`` check at ``isSendmail()`` function works fine because, into PHPMailer side of view, it's calling the sendmail executable.

But packages like this are not available to all distros - CentOS didn't offer one. So if I want to use the ``mail()`` function with PHPMailer, I need to replace sendmail executable directly through the following commands (CentOS 8):

```
cd /etc/alternatives/
mv mta mta.orig
mv mta-sendmail mta-sendmail.orig
ln -s /usr/bin/msmtp mta
ln -s /usr/bin/msmtp mta-sendmail
```

So I finally could use ``msmtp`` with PHPMailer, but the ``sendmail_path`` directive should be respected instead.

# Code to reproduce

# Debug output

**Comment by Synchro** [comment]
I think you're confusing PHP's built-in sending mechanism (`mail()`) with PHPMailer sendmail function. The `sendmail_path` ini value is used directly by the mail function, so you can do this:

    ini_set('sendmail_path', '/path/to/msmtp');

and then when you send using `mail()` (including via PHPMailer), it will use your path. The sendmail option in PHPMailer is there as most sendmail-compatible MTAs (e.g. qmail & postfix) provide an explicit replacement for the default `sendmail` binary, and that's what PHPMailer is expecting to see.

[I wrote a post about using this mechanism with HELO](https://usehelo.com/blog/how-to-use-helo-with-phps-mail-function) on their blog.

I can see from the history that I added this check back in 2009, though I don't recall the exact reason for doing so! It was possibly because of support questions created by people pointing it at something that wasn't a sendmail-compatible binary.

As an alternative, you could subclass PHPMailer and override that method; I can see that working well in this case.

**Comment by coldscientist** [comment]
Hello Synchro, thank you for the faster response!
The problem is that GLPI uses PHPMailer to send mails, it offers to use "PHPMailer" or "Mail" to send mails, but if you analyse GLPI code, you'll see that it uses PHPMailer to wrap the mail code, so it doesn't use the mail PHP function itself, and PHPMailer is always called. PHPMailer checks the word "sendmail" into sendmail_path PHP directive through isSendmail() function and if the word "sendmail" is not found (aka. Using a fake MTA, like msmtp or ssmtp), it overrides sendmail_path PHP directive instead of using the original directive. So it doesn't allow using custom executables (like msmtp) directly, so msmtp works when called through mail() function directly, but it is overridden by sendmail executable when called through PHPMailer because PHPMailer overrides it with sendmail through isSendmail() function.
The problem is that I do not develop GLPI on my own and I can't change source code to use sendmail_path directive. I understand your point and I believed that it was the main reason for isSendmail() function exists, but it's difficult to debug the reason because GLPI mail function wasn't working because this behavior isn't documented at GLPI and/or PHPMailer troubleshooting guide. I believe that this override should be cited at PHPMailer troubleshooting guide so the users would know that PHPMailer ignores sendmail_path directive with custom executables and the sendmail executable should be replaced instead of changing this PHP directive, but I believe that this PHP directive should be respected, because if it was set manually, I want to believe that the user that edited it knows what he/she was doing and if something stopped working, he/she knows that the directive was edited and that this edit can be the root cause of any trouble.

**Comment by Synchro** [comment]
You're missing something here: PHPMailer's default behaviour is to use the `mail()` function to do the sending, so the choice isn't between PHPMailer and mail(), it's whether PHPMailer uses mail() or another mechanism. You don't need to use `isSendmail()` at all.

**Comment by coldscientist** [comment]
Hello Synchro, thank you for your response! It was my fault!
I believe that PHPMailer called isSendmail() function internally before calling mail() function. But neither PHPMailer or GLPI calls it. I figure it out that GLPI mail were being sent when I replaced sendmail executable with msmtp, so I decided to see the sendmail_path that PHP was using, that was slightly different than the one I was using based on [Arch Linux wiki](https://wiki.archlinux.org/index.php/msmtp).

The one based at Arch Linux wiki:
```
sendmail_path = "/bin/msmtp -C /etc/msmtprc -t  --read-envelope-from"
```

The one that PHP was using by default (and that was working when I replaced sendmail binaries with msmtp ones):
```
sendmail_path = "/usr/sbin/sendmail -t -i -f webmaster@glpi.example.com"
```

So I decided to replace only the 'sendmail' path with 'msmtp' path instead. My final php.ini was:
```
sendmail_path = "/usr/bin/msmtp -t -i -f webmaster@glpi.sulcromo.com.br"
```

And it worked! The e-mails was being sent nicely through msmtp using PHPMailer and GLPI!

Thank you for your help Synchro! Because of your comments, I see that the problem wasn't PHPMailer or GLPI itself (they didn't call the isSendmail() function at all), but it was somehow related to the configuration I was using at sendmail_path directive at PHP. I will close this 'issue' ticket, hope our discussion helps someone in the future.
